Minor corrections with empty arrays (#3087)
modelica/ModelicaSpecification: chapters/arrays.tex

Real A[0, 3], B[5, 0], C[0, 0]; // empty matrices
\end{lstlisting}

> Empty matrices can be constructed using the \lstinline!fill! function.  For example:

Reviewer: Switch to sentence-based line breaks when making edits:
```suggestion
Empty matrices can be constructed using the \lstinline!fill! function.
For example:
```

> Empty matrices can be constructed with the fill function.  For example:

Reviewer: We should take the opportunity to also fix the formatting here (also _using_ sounds better to me than _with_):
```suggestion
Empty matrices can be constructed using the \lstinline!fill! function.  For example:
```

Reviewer 2: Agreed.

\begin{lstlisting}[language=modelica]

> Real A[:,:] = fill(0.0, 0, 1); // a Real 0 x 1 matrix

Reviewer 2: I actually prefer the original here, as adding _empty_ to me suggests that there could be _non-empty_ 0 x 1 matrices.
Since it is in the section about Empty Arrays I don't think we need to be that explicit that it is empty.

Reviewer: But shouldn't we remove "empty" from all comments then?

Reviewer 2: Not the ones that say "empty matrices" (of different dimensions) and "empty vector" that don't specify the size.
What I don't like is the redundancy of writing empty _and_ writing an empty size.

Author: OK.

> Boolean B[:, :, :] = fill(false, 0, 1, 0); // an empty Boolean 0 x 1 x 0 matrix

Reviewer: Did you verify that the source code line doesn't wrap?

Author: No, I did not.

\end{lstlisting}

> It is not possible to access the elements of an empty matrix, e.g.\ \lstinline!v[j, k]! cannot be evaluated if \lstinline!v = []! because the assertion fails

Reviewer 2: I actually prefer 'an element' to 'the elements'. The reason is that `v[:,:]` can be seen as accessing the elements of the matrix, but `v[j,k]` for scalar `j` and `k` accesses _an_ element.
```suggestion
It is not possible to access an element of an empty matrix, e.g.\ \lstinline!v[j, k]! for scalars \lstinline!j! and \lstinline!k! cannot be evaluated if \lstinline!v = []! because the assertion fails
```

Reviewer 3: It should also be merged with the next line because the next line is part of the same sentence. Also "the assertion fails that..." is not quite correct. It is better to say "the assertion that... fails."

Reviewer: Note my suggestion to replace this sentence with more useful examples is taking care of this.

Reviewer 3: Yes, there is also @HansOlsson's suggestion above which does not correct the line break and the weird phrasing. So I was not sure which of the suggestions is considered as the main candidate.

> It is not possible to access the elements of an empty matrix, e.g.\ \lstinline!v[j, k]! cannot be evaluated if \lstinline!v = []! because the assertion fails
> that the indices must be greater than zero.

Reviewer: It would probably be better to rewrite the sentence completely, as there are no _scalar_ elements in an empty matrix/array.  However, by not specifying all subscripts the array may actually be both empty and indexable at the same time.  Maybe it's more interesting to illustrate this point?
```suggestion
\begin{example}
Whereas indexing into an empty dimension of an array is an error, not all applications of indices to empty arrays are invalid:
\begin{lstlisting}
Real[1, 0] a = fill(0.0, 1, 0); // empty 1 x 0 matrix
Real[0] a1 = a[1]; // empty vector
\end{lstlisting}
\end{example}
```

(Note: The suggestion above was written without testing it.)

Reviewer 2: I think the more common and important case would be `:` and range-indexing; such as `a[1, :]` and `a[1, 1:end]`.

We could use the terminology of 10.5 (Array) Indexing and say that scalar access is not allowed as there are no scalar elements, but slicing generating arrays can be legal.

Reviewer: I edited the suggestion above to also include slicing with `:`.

Reviewer 3: I like @henrikt-ma's suggestion and the examples but I also like that the PR version explicitly says that indices must be greater than zero. I find it more clear. It might be  Another thing in @henrikt-ma's suggestion that I am wondering about is that the text is now part of the example and would appear formatted as a nonnormative text. The original text is part of the formal text. So I am wondering if it should remain as formal text. Or is it OK to lessen the strength/importance of that statement?

Reviewer: The upper and lower index bounds are given in https://specification.modelica.org/master/arrays.html#array-dimension-lower-and-upper-index-bounds, so we shouldn't repeat this normatively here.

Multiplication of two empty matrices results in a zero matrix of corresponding numeric type if the result matrix has no zero dimension sizes, i.e.,
\begin{lstlisting}[language=modelica]
Real[0, m] * Real[m, n] = Real[0, n] // empty matrix
Real[m, n] * Real[n, 0] = Real[m, 0] // empty matrix

> Real[m, 0] * Real[0, n] = fill(0.0, m, n) // non-empty matrix, with zero elements.

Reviewer: Consistent use of punctuation and reformulating to avoid ambiguity:
```suggestion
Real[m, 0] * Real[0, n] = fill(0.0, m, n) // non-empty matrix of zeros
```
